### PR TITLE
replication,cmd: improve flavor handling

### LINKED ISSUE #944: go-mysqlbinlog: wrong positions for events and Gtid events get decoded as QueryEvent
Server: MariaDB 11.4.2

From `go-mysqlbinlog`:
```
=== RotateEvent ===
Date: 1970-01-01 01:00:00
Log position: 0
Event size: 48
Position: 4
Next log name: mysqld-bin.000002

=== FormatDescriptionEvent ===
Date: 2024-11-08 08:55:18
Log position: 256
Event size: 252
Version: 4
Server version: 11.4.2-MariaDB-ubu2404-log
Checksum algorithm: 1

=== UserVarEvent ===
Date: 2024-11-08 08:55:18
Log position: 285
Event size: 29
Event data: 
00000000  01 00 00 00 21 01                                 |....!.|

=== QueryEvent ===
Date: 2024-11-08 08:55:18
Log position: 329
Event size: 44
Slave proxy ID: 0
Execution time: 0
Error code: 0
Schema: 
Query: # Dummy

=== QueryEvent ===
Date: 2024-11-08 09:24:40
Log position: 371
Event size: 42
Slave proxy ID: 0
Execution time: 0
Error code: 0
Schema: 
Query: # Dum

=== QueryEvent ===
Date: 2024-11-08 09:24:40
Log position: 458
Event size: 87
Slave proxy ID: 11
Execution time: 0
Error code: 0
Schema: test
Query: create database test

=== QueryEvent ===
Date: 2024-11-08 09:27:31
Log position: 500
Event size: 42
Slave proxy ID: 0
Execution time: 0
Error code: 0
Schema: 
Query: # Dum

=== QueryEvent ===
Date: 2024-11-08 09:27:31
Log position: 715
Event size: 215
Slave proxy ID: 11
Execution time: 0
Error code: 0
Schema: test
Query: create table test_time(
           id int(11) primary key, 
         `time_1` time DEFAULT '-838:59:59' COMMENT '-838:59:59 to 838:59:59'
)
```

From the client:
```
mysql-11.4.2-MariaDB-ubu2404-log> show binlog events in 'mysqld-bin.000002';
+-------------------+-----+-------------------+-----------+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| Log_name          | Pos | Event_type        | Server_id | End_log_pos | Info                                                                                                                                                    |
+-------------------+-----+-------------------+-----------+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| mysqld-bin.000002 |   4 | Format_desc       |         1 |         256 | Server ver: 11.4.2-MariaDB-ubu2404-log, Binlog ver: 4                                                                                                   |
| mysqld-bin.000002 | 256 | Gtid_list         |         1 |         285 | []                                                                                                                                                      |
| mysqld-bin.000002 | 285 | Binlog_checkpoint |         1 |         329 | mysqld-bin.000002                                                                                                                                       |
| mysqld-bin.000002 | 329 | Gtid              |         1 |         371 | GTID 0-1-1                                                                                                                                              |
| mysqld-bin.000002 | 371 | Query             |         1 |         458 | create database test                                                                                                                                    |
| mysqld-bin.000002 | 458 | Gtid              |         1 |         500 | GTID 0-1-2                                                                                                                                              |
| mysqld-bin.000002 | 500 | Query             |         1 |         715 | use `test`; create table test_time(
           id int(11) primary key, 
         `time_1` time DEFAULT '-838:59:59' COMMENT '-838:59:59 to 838:59:59'
) |
+-------------------+-----+-------------------+-----------+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
7 rows in set (0.00 sec)
```

Note this event:
```
=== QueryEvent ===
Date: 2024-11-08 09:27:31
Log position: 500
Event size: 42
Slave proxy ID: 0
Execution time: 0
Error code: 0
Schema: 
Query: # Dum
```

```
mysql-11.4.2-MariaDB-ubu2404-log> SHOW BINLOG EVENTS IN 'mysqld-bin.000002' FROM 458 LIMIT 1\G
*************************** 1. row ***************************
   Log_name: mysqld-bin.000002
        Pos: 458
 Event_type: Gtid
  Server_id: 1
End_log_pos: 500
       Info: GTID 0-1-2
1 row in set (0.00 sec)
```

Issues:
1. `Log position: 500`: This is not the position of the event as that should be 458. It is the end position of the event.
2. This is a `Gtid` event and not a `QueryEvent`
3. `Query: # Dum`: I think this should have said "Dummy", but that got truncated. But this should probably have had `GTID 0-1-2` instead.

Issues with other events:
4. `Gtid_list` is decoded as `UserVarEvent`.
5. `Binlog_checkpoint` is decoded as `QueryEvent`

### PR DESCRIPTION
**Pull request by dveeden**
Issue: close #944 

For `go-mysqlbinlog` and `go-canal`: return an error for invalid flavors.

For `go-mysqlbinlog`: log a non-critical error if the server version has "MariaDB" in it but the flavor isn't set to "mariadb".

### CODE REVIEW

**Comment by dveeden** [comment]
If there is an good way to make it auto detect the flavor this would be even better

File: replication/binlogsyncer.go [hunk 1]
```diff
@@ -424,6 +425,15 @@ func (b *BinlogSyncer) GetNextPosition() Position {
 	return b.nextPos
 }
 
+func (b *BinlogSyncer) checkFlavor() {
+	serverVersion := b.c.GetServerVersion()
+	if b.cfg.Flavor != MariaDBFlavor &&
+		strings.Contains(b.c.GetServerVersion(), "MariaDB") {
```
Comment (lance6716) [inline replication/binlogsyncer.go:431]: The server returns `MariaDB` but our flavor requires lowercased, this is a bit uncomfortable. I orefer we let `ValidateFlavor` `ToLower` the string before compare.

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: replication/binlogsyncer.go [hunk 2]
```diff
@@ -424,6 +425,15 @@ func (b *BinlogSyncer) GetNextPosition() Position {
...
+		b.cfg.Logger.Errorf("misconfigured flavor (%s) for server %s",
```
Comment (lance6716) [inline replication/binlogsyncer.go:432]: And I think we can add comment to explain why this is only a log not a failure to caller, like "MariaDB still sends binlog events in a compatible way although some functionalities are lost, so we just log the problem"

File: replication/binlogsyncer.go (hunk 1 above)
Comment (dveeden) [reply replication/binlogsyncer.go:431]: I changed `ValidateFlavor()` to do the `ToLower()` as suggested

File: replication/binlogsyncer.go (hunk 2 above)
Comment (dveeden) [reply replication/binlogsyncer.go:435]: I've added a code comment. Or did you want to put this in the message instead?

**Review by lance6716 — APPROVED** [review]

**Pull request merged**